// botframework-webchat-component/lib/ScreenReaderText.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyleToEmotionObject = _interopRequireDefault(require("./hooks/internal/useStyleToEmotionObject"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint react/forbid-dom-props: ["off"] */
var ROOT_STYLE = {
  // .sr-only - This component is intended to be invisible to the visual Web Chat user, but read by the AT when using a screen reader
  color: 'transparent',
  height: 1,
  overflow: 'hidden',
  position: 'absolute',
  // We need to set top: 0, otherwise, it will repro:
  // - Run NVDA
  // - Make the transcript long enough to show the scrollbar
  // - Press SHIFT-TAB, focus on upload button
  // - Press up arrow multiple times
  top: 0,
  whiteSpace: 'nowrap',
  width: 1
};

var ScreenReaderText = function ScreenReaderText(_ref) {
  var id = _ref.id,
      text = _ref.text;
  var rootClassName = (0, _useStyleToEmotionObject.default)()(ROOT_STYLE) + '';
  return /*#__PURE__*/_react.default.createElement("div", {
    className: rootClassName,
    id: id
  }, text);
};

ScreenReaderText.defaultProps = {
  id: undefined
};
ScreenReaderText.propTypes = {
  id: _propTypes.default.string,
  text: _propTypes.default.string.isRequired
};
var _default = ScreenReaderText;
exports.default = _default;
