// botframework-webchat-api/lib/hooks/middleware/createDefaultGroupActivitiesMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createDefaultGroupActivityMiddleware;

var _botframeworkWebchatCore = require("botframework-webchat-core");

var SENT = _botframeworkWebchatCore.Constants.ActivityClientState.SENT;

function bin(items, grouping) {
  var lastBin;
  var bins = [];
  var lastItem;
  items.forEach(function (item) {
    if (lastItem && grouping(lastItem, item)) {
      lastBin.push(item);
    } else {
      lastBin = [item];
      bins.push(lastBin);
    }

    lastItem = item;
  });
  return bins;
}

function sending(activity) {
  return activity.from.role === 'user' && activity.channelData && activity.channelData.state !== SENT;
}

function shouldGroupTimestamp(activityX, activityY, groupTimestamp) {
  if (groupTimestamp === false) {
    // Hide timestamp for all activities.
    return true;
  } else if (activityX && activityY) {
    if (sending(activityX) !== sending(activityY)) {
      return false;
    }

    groupTimestamp = typeof groupTimestamp === 'number' ? groupTimestamp : Infinity;
    var timeX = new Date(activityX.timestamp).getTime();
    var timeY = new Date(activityY.timestamp).getTime();
    return Math.abs(timeX - timeY) <= groupTimestamp;
  }

  return false;
}

function createDefaultGroupActivityMiddleware(_ref) {
  var groupTimestamp = _ref.groupTimestamp;
  return function () {
    return function () {
      return function (_ref2) {
        var activities = _ref2.activities;
        return {
          sender: bin(activities, function (x, y) {
            return x.from.role === y.from.role;
          }),
          status: bin(activities, function (x, y) {
            return shouldGroupTimestamp(x, y, groupTimestamp);
          })
        };
      };
    };
  };
}
