// botframework-webchat-api/lib/hooks/useCreateActivityStatusRenderer.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useCreateActivityStatusRenderer;

var _botframeworkWebchatCore = require("botframework-webchat-core");

var _react = _interopRequireWildcard(require("react"));

var _useGetSendTimeoutForActivity = _interopRequireDefault(require("./useGetSendTimeoutForActivity"));

var _useTimePassed = _interopRequireDefault(require("./internal/useTimePassed"));

var _useWebChatAPIContext2 = _interopRequireDefault(require("./internal/useWebChatAPIContext"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* eslint react/prop-types: "off"*/
var _Constants$ActivityCl = _botframeworkWebchatCore.Constants.ActivityClientState,
    SEND_FAILED = _Constants$ActivityCl.SEND_FAILED,
    SENDING = _Constants$ActivityCl.SENDING,
    SENT = _Constants$ActivityCl.SENT;

var ActivityStatusContainer = function ActivityStatusContainer(_ref) {
  var activity = _ref.activity,
      hideTimestamp = _ref.hideTimestamp,
      nextVisibleActivity = _ref.nextVisibleActivity;

  var _useWebChatAPIContext = (0, _useWebChatAPIContext2.default)(),
      createActivityStatusRenderer = _useWebChatAPIContext.activityStatusRenderer;

  var getSendTimeoutForActivity = (0, _useGetSendTimeoutForActivity.default)(); // SEND_FAILED from the activity is ignored, and is instead based on styleOptions.sendTimeout.
  // Note that the derived state is time-sensitive. The useTimePassed() hook is used to make sure it changes over time.

  var _activity$channelData = activity.channelData;
  _activity$channelData = _activity$channelData === void 0 ? {} : _activity$channelData;
  var _activity$channelData2 = _activity$channelData.clientTimestamp,
      clientTimestamp = _activity$channelData2 === void 0 ? 0 : _activity$channelData2,
      state = _activity$channelData.state,
      role = activity.from.role;
  var activitySent = state !== SENDING && state !== SEND_FAILED;
  var fromUser = role === 'user';
  var sendTimeout = getSendTimeoutForActivity({
    activity: activity
  });
  var pastTimeout = (0, _useTimePassed.default)(fromUser && !activitySent ? new Date(clientTimestamp).getTime() + sendTimeout : 0);
  var sendState = activitySent || !fromUser ? SENT : pastTimeout ? SEND_FAILED : SENDING;
  return (0, _react.useMemo)(function () {
    return createActivityStatusRenderer({
      activity: activity,
      hideTimestamp: hideTimestamp,
      nextVisibleActivity: nextVisibleActivity,
      // "nextVisibleActivity" is for backward compatibility, please remove this line on or after 2022-07-22.
      sameTimestampGroup: hideTimestamp,
      // "sameTimestampGroup" is for backward compatibility, please remove this line on or after 2022-07-22.
      sendState: sendState
    });
  }, [activity, createActivityStatusRenderer, hideTimestamp, nextVisibleActivity, sendState]);
};

function useCreateActivityStatusRenderer() {
  return (0, _react.useMemo)(function () {
    return function (_ref2) {
      var activity = _ref2.activity,
          nextVisibleActivity = _ref2.nextVisibleActivity;
      return function () {
        var _ref3 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
            hideTimestamp = _ref3.hideTimestamp;

        return /*#__PURE__*/_react.default.createElement(ActivityStatusContainer, {
          activity: activity,
          hideTimestamp: hideTimestamp,
          nextVisibleActivity: nextVisibleActivity
        });
      };
    };
  }, []);
}
