// botframework-webchat-component/lib/Styles/StyleSet/SuggestedActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSuggestedActionsStyle;

/* eslint no-empty-pattern: "off" */

/* eslint no-magic-numbers: ["error", { "ignore": [2] }] */
function createSuggestedActionsStyle(_ref) {
  var paddingRegular = _ref.paddingRegular,
      suggestedActionsStackedHeight = _ref.suggestedActionsStackedHeight,
      suggestedActionsStackedOverflow = _ref.suggestedActionsStackedOverflow,
      transcriptOverlayButtonBackground = _ref.transcriptOverlayButtonBackground,
      transcriptOverlayButtonBackgroundOnDisabled = _ref.transcriptOverlayButtonBackgroundOnDisabled,
      transcriptOverlayButtonBackgroundOnFocus = _ref.transcriptOverlayButtonBackgroundOnFocus,
      transcriptOverlayButtonBackgroundOnHover = _ref.transcriptOverlayButtonBackgroundOnHover,
      transcriptOverlayButtonColor = _ref.transcriptOverlayButtonColor,
      transcriptOverlayButtonColorOnDisabled = _ref.transcriptOverlayButtonColorOnDisabled,
      transcriptOverlayButtonColorOnFocus = _ref.transcriptOverlayButtonColorOnFocus,
      transcriptOverlayButtonColorOnHover = _ref.transcriptOverlayButtonColorOnHover;
  return {
    '&.webchat__suggested-actions': {
      '& .webchat__suggested-actions__carousel': {
        paddingBottom: paddingRegular / 2,
        paddingTop: paddingRegular / 2,
        '& .react-film__filmstrip': {
          scrollbarWidth: 'none',
          '& .react-film__filmstrip__item:first-child': {
            paddingLeft: paddingRegular / 2
          },
          '& .react-film__filmstrip__item:last-child': {
            paddingRight: paddingRegular / 2
          }
        },
        '& .react-film__flipper': {
          '&:disabled, &[aria-disabled="true"]': {
            '& .react-film__flipper__body': {
              backgroundColor: transcriptOverlayButtonBackgroundOnDisabled,
              color: transcriptOverlayButtonColorOnDisabled
            }
          },
          '&:focus .react-film__flipper__body': {
            backgroundColor: transcriptOverlayButtonBackgroundOnFocus,
            color: transcriptOverlayButtonColorOnFocus || transcriptOverlayButtonColor
          },
          '&:hover .react-film__flipper__body': {
            backgroundColor: transcriptOverlayButtonBackgroundOnHover,
            color: transcriptOverlayButtonColorOnHover || transcriptOverlayButtonColor
          },
          '& .react-film__flipper__body': {
            background: transcriptOverlayButtonBackground,
            color: transcriptOverlayButtonColor,
            outline: 0
          }
        }
      },
      '& .webchat__suggested-actions__stack': {
        maxHeight: suggestedActionsStackedHeight || 'auto',
        overflowY: suggestedActionsStackedOverflow || 'auto',
        paddingBottom: paddingRegular / 2,
        paddingLeft: paddingRegular / 2,
        paddingRight: paddingRegular / 2,
        paddingTop: paddingRegular / 2
      }
    }
  };
}
