// botframework-webchat-component/lib/Styles/StyleSet/WarningNotification.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createWarningNotificationStyle;

function createWarningNotificationStyle(_ref) {
  var connectivityIconPadding = _ref.connectivityIconPadding,
      connectivityMarginLeftRight = _ref.connectivityMarginLeftRight,
      connectivityMarginTopBottom = _ref.connectivityMarginTopBottom,
      connectivityTextSize = _ref.connectivityTextSize,
      primaryFont = _ref.primaryFont,
      slowConnectivity = _ref.slowConnectivity,
      notificationText = _ref.notificationText;
  return {
    alignItems: 'center',
    color: notificationText,
    // TODO: [P2] We should not set "display" in styleSet, this will allow the user to break the layout for no good reasons.
    display: 'flex',
    fontFamily: primaryFont,
    fontSize: connectivityTextSize,
    marginBottom: connectivityMarginTopBottom,
    marginLeft: connectivityMarginLeftRight,
    marginRight: connectivityMarginLeftRight,
    marginTop: connectivityMarginTopBottom,
    '& > svg': {
      fill: slowConnectivity,
      '&:not(webchat__warning--rtl)': {
        paddingRight: connectivityIconPadding
      },
      '&.webchat__warning--rtl': {
        paddingLeft: connectivityIconPadding
      }
    }
  };
}
